envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/snapshot/SnapshotUpdater.kt

```diff
@@ -9,19 +9,20 @@ import pl.allegro.tech.servicemesh.envoycontrol.services.LocalityAwareServicesSt
 import reactor.core.publisher.Flux
 import reactor.core.publisher.Mono
 import reactor.core.scheduler.Scheduler
-import java.util.function.BiFunction
 
 class SnapshotUpdater(
     private val cache: SnapshotCache<Group>,
     private val properties: SnapshotProperties,
     private val scheduler: Scheduler,
-    private val onGroupAdded: Flux<out Any>,
+    private val onGroupAdded: Flux<List<Group>>,
     private val meterRegistry: MeterRegistry
 ) {
     companion object {
         private val logger by logger()
     }
 
+    private lateinit var lastAdsSnapshot: Snapshot
```

Comment:
We only re-assign this variable to another Snapshot, we never modify the Snapshot itself. And our Fluxes aren't running in parallel mode. So I think we are safe in terms of concurrent modification and race conditions on snapshot re-assigning.

But there still may be a problem with "visibility" of the latest Snapshot to all threads. I think `@Volatile` annotation would be enough to prevent this problem.

Reply:
I'm not convinced by this approach. Also, volatile is not enough as far as I understand. It helps with visibility, but updates can lead to corrupt references. Consult this SO thread: https://stackoverflow.com/questions/2576513/is-writing-a-reference-atomic-on-64bit-vms

Reply:
> Also, volatile is not enough as far as I understand. It helps with visibility, but updates can lead to corrupt references.

I do not understand this at all. The SO thread you linked to says: "Writes to and reads of references are always atomic, regardless of whether they are implemented as 32-bit or 64-bit values." so how can this lead to corrupt references? (even more so, now that there is @Volatile)

Reply:
@chemicL From the thread you linked:
 
> Writes and reads of volatile long and double values are always atomic.
Writes to and reads of references are always atomic, regardless of whether they are implemented as 32-bit or 64-bit values.

> Implementations of the Java Virtual Machine are encouraged to avoid splitting 64-bit values where possible. Programmers are encouraged to declare shared 64-bit values as **volatile** or synchronize their programs correctly to avoid possible complications.

As long as our code, that modifies this reference is sequential (it is right now), I think the visibility is our only concern and @Volatile is enough.

> I'm not convinced by this stateful approach.
Me too, but I don't have an idea right now how to do it in "fluxy" way.

Reply:
> Also, volatile is not enough as far as I understand. It helps with visibility, but updates can lead to corrupt references.

I do not understand this at all. The SO thread you linked to says: "Writes to and reads of references are always atomic, regardless of whether they are implemented as 32-bit or 64-bit values." so how can this lead to corrupt references? (even more so, now that there is @Volatile)

Reply:
Also: https://stackoverflow.com/questions/281132/java-volatile-reference-vs-atomicreference

Reply:
Right, I wasn't reading carefully it seems. I focused on the part about splitting writes for longs and doubles. Thanks for pointing this out.
Idea: if we are to stay with reactor here - perhaps merging a flux that does a cache emit of the latest value and for each new emitted group we are presented with the latest stored snapshot instead of this paradigm mix?

Reply:
I updated the version and it now uses `scan` to cache the latest value. Check out if this is more acceptable.

Reply:
Thanks, looks I like this approach. The separation of concerns (synchronisation mechanics) will be beneficial in the future.